### ISSUE TITLE
pip install tries to upgrade django to 1.7

### ISSUE THREAD

**Issue by BrunoDesthuilliers** [django-debug-toolbar/django-debug-toolbar#766]
In a django 1.6 project - virtualenv with django already installed -, `pip install django-debug-toolbar` tries to install django >= 1.7.

I understand that current version requires django >= 1.7 so it's the "expected" behaviour *but* as far as I'm concerned a django app forcing django version update is neither expected nor  right - it should either ask for confirmation or refuse to install with a clear message if a django version < 1.7 is already installed (don't know wether pip / setuptools etc allow for such a behaviour tough).

**Comment by aaugustin** [comment]
Can you submit a patch?

**Comment by BrunoDesthuilliers** [comment]
I would have if I was a bit more fluent with python packaging (very limited experience here sorry).... As I mentionned, I don't even know if it's actually technically possible to do something like this. 

But anyway : the best fix AMHA would be to plain remove the requirement on django 1.7 and clearly document which django version is supported (I know the information is there - I found it browsing the tags on github and finally installed 1.3.2 - but it's not obvious on readthedoc).  That's what most django apps do FWIW, debug_toolbar is the first one I've met that tried to update my django install. 

My 2 cents...

**Comment by aaugustin** [comment]
The dependency seems to have been added as a side-effect of an unrelated change in https://github.com/django-debug-toolbar/django-debug-toolbar/commit/eacf116d61f0e3f47a7c581275a18b3aec8548b1#diff-2eeaed663bd0d25b7e608891384b7298L15.

I'm OK with dropping it.

**Comment by BrunoDesthuilliers** [comment]
pull request submitted, build ok.

**Comment by vstoykov** [comment]
I don't see any problem here. In the setup the line telling that Django debug toolbar want's Django >= 1.7 must stay. This is the dependency. Pip is clever and try to install dependencies (for that reason we use PIP and not downloading packages manually and executing python setup.py install on them).

@BrunoDesthuilliers If your project is not ready for django 1.7 (which is current minor supported version not only from django-debug-toolbar but from Django itself) then you need to use requiements.txt with your packages and versions to which it depends and to put `django-debug-toolbar==1.3.2` because this is your dependency.

I will vote this issue to be marked as "invalid".

**Comment by BrunoDesthuilliers** [comment]
@vstoykov  I'm sorry but I have to disagree here. An add-on should in no way try to upgrade the host framework's version - which is the responsability of the project's maintainer - specially when this add-on is supposed to be a developper-only tool (I could eventually see the point for a full-blown CMS or E-Commerce application and even then...). FWIW django-debug-toolbar is the very first django app I've seen that behave that way - the others only document which django version each release supports, and break if you install them on an unsupported one, and as a Django dev that's what I expect.  

And yes, switching to 1.3.2 is what I did (obviously). The issue here is not with the current django-debug-toolbar version not supporting django 1.6 but with it wanting to upgrade django not caring if it breaks my whole project.

**Comment by vstoykov** [comment]
django-debug-toolbar do not try to install newer version of Django. PIP is installing it because it sees that this is requirement which is not satisfied (this is how PIP works).

When you install package and then when you run your project something fails you start wondering where is the problem - in your code, in the 3th party app code or something else. You can submit an issue because you think that something is broken until you realize that you are trying to use unsupported version.

On the other hand when you install a package and it tries to upgrade some other package (mainly Django) you now know that obviously you are trying to use latest version of something which do not support your older version of Django.

Your not seeing many (or other) packages to depends on Django 1.7 or greater right now because a lot of 3th party packages are trying to achieve greater backward compatibility and support Django versions to 1.4 or even lower.

Probably the main problem is that support for lower versions of Django was dropped so soon (only 5 days after end of life of Django 1.4 the previous LTS), but the approach how it was done was right.

A proper solution for the situation is probably more precise documentation about installing `django-debug-toolbar` on different versions, and advice for upgrading to newer Django.

@BrunoDesthuilliers  I have one suggestion to you. Youse dev_requirements.txt or requirements_dev.txt or something similar to describe all dependencies not needed for production bot only for development. This will be easier for you and your colleagues.

@aaugustin I have one suggestion to you also. Think more before dropping support of older Django versions. I don't say that all Django versions from the beginning of time need to be supported, but still only 5 days after end of life it's to soon.

**Comment by BrunoDesthuilliers** [comment]
@vstoykov  I have one suggestion for you too : don't assume that peoples who disagree with you and state they are not packaging experts are noobies needing guidance ;)  Oh and I didn't wait for your "suggestion" to separate dev requirements from prod's requirements either - believe it or else but it's not exactly my first Python project (nor the first Django one).

The code stating that latest django-debug-toolbar requires django >= 1.7 is in django-debug-toolbar's code, so yes practically speaking it actually is django-debug-toolbar that asks pip to upgrade django.  

You can sample other django reusable apps - I just had a look at a couple ones from one of my project's requirements - and very few (actually 2 out of 8) mention Django in the the setup.py `requires` or `install_requires`. Not enough for statistical evidence tough...   Now wether it's a "good practice" or not is debatable but as far as I'm concerned I fail to see how someone installing a django app in a project's venv would expect the app to work without django anyway. What I do not know is if there's a way to tell pip to refuse to install or issue a warning instead, which would make much more sense.

This being said you might be right about the point that the root problem here is dropping support for older Django versions a tad too soon **and** not clearly documenting supported versions.

**Comment by aaugustin** [comment]
@vstoykov I've spent days on end discussing and thinking about how to handle backwards incompatible changes and version upgrades in Django and its ecosystem. Please don't assume that I made this change on a whim.

When a Django version no longer receives security support, it's very damaging to the community to maintain support for it in a popular application such as the debug-toolbar because it implies that it's reasonable to keep running it and that implication comes from people who are supposed to know what they are doing. (That's just one reason)

----

Let's get back to the original topic.

In general pluggable Django applications don't make Django a requirement because it's fairly obvious that they only work with Django. There's no use case for doing `pip install django-debug-toolbar` and exepecting it to install (or upgrade) Django.

The desirable behavior here is to warn the user if an incompatible version of Django is installed. Forcing the upgrade of Django is one way to warn but perhaps not the best one. Aborting the installation would work as well. Silently proceeding sounds bad -- because the app will crash immediately.

**Comment by vstoykov** [comment]
@BrunoDesthuilliers Ok I accepts your advice and I will try to not judge people too early :).

About listing django as requirement in pluggable django applications if it says only `Django` yes it's obvious and unnecessary, but adding version is good approach. It explicitly says which versions are supported.

For developer executing  `pip install django-debug-toolbar` it's obvious that he/she do not want to upgrade Django right now but when this was happening this means that the code base is really out of date and eventually insecure and probably now is the time to think for upgrade. 

There always a way for a workaround and print some warning on screen or even break but:
 1. If you missed that your django version are updated automatically, probably you will miss the warning message also.
 2. If install process breaks because of dependancy, why we use `setuptools` and `pip` and not `distutils`.

Probably for the current situation there is no "best" solution because the problem is not in `django-debug-toolbar` setup but in handling python packages at general. How given package which as plugin to bigger system can describe his compatibility and package manager (pip or easy_install) to handle this accordingly. This is not related to Django ecosystem but to all python packages in general.


P.S.

Now most of the django plugins probably do not include requirement to Django because they do not want to mess with broken situation in Python packagin.

P.S.

@BrunoDesthuilliers @aaugustin Sorry if my words are offending you or I sound little more rude. It was not intentional. English is not my native language and probably some times I use inappropriate words.

**Comment by aaugustin** [comment]
@vstoykov No problem, I guessed so :-)

**Comment by BrunoDesthuilliers** [comment]
@vstoykov  no offense, and English is not my native language either so I can well understand how difficult it can be to convey the exact intention sometimes ;)

@aaugustin @vstoykov Yes indeed there might be something about Python's packaging ecosystem that doesn't quite cut it when it comes to "plugin" packages - as I said I'm far from an expert at Python packaging -, and if then yes that's a more general issue and no there's no obvious best solution here.

As a side note: one shouldn't assume that it's the developper that needs to be reminded that upgrading to a supported & more secure Django release would be a good thing - when the stakeholder don't want to pay for an upgrade despite your warnings, then you don't upgrade, period. 

Now thinking about the whole thing: if the readme contained a clear mention of which Django version is required for the current debug-toolbar release I would not have tried a pip install on this release so I would not have been surprised by the install trying to upgrade Django.  

I've been actually surprised yesterday when I found out that at least a couple other django apps I use had requirements on Django version and then wondered why I never had any issue with this before (and I've been using Django and virtualenv for years on tens of projects).  Then I remembered how many times I've been thru the process of read the readme, notice the Django version requirement, scroll thru the past releases to find the latest one matching my project's django version and explicitely install this one. 

To make a long story short; at least part of the solution might be more on the doc side than on the technical one ;)
